### PR TITLE
Db stuff

### PR DESCRIPTION
**Pull request by manfredlift**
Initial db stuff.

Run with a temp db: `python -m parking.backend --test_db`
Run with a real db: `python -m parking.backend --db <db_url>`

### CODE REVIEW

File: parking/backend/__main__.py [hunk 1]
```diff
@@ -1,13 +1,34 @@
...
+            _main(postgresql.url())
+    else:
+        _main(db_url)
 
 
 if __name__ == "__main__":
-    main()
+    parser = argparse.ArgumentParser(description='Parking backend.')
+    parser.add_argument("--db", help="Database full url.", default="postgresql://localhost/postgres")
+    parser.add_argument("--test_db", action='store_true', help="Create a temporary database.")
```
Comment (HarveyHunt) [inline parking/backend/__main__.py]: Maybe this could be hidden behind a generic ```--debug``` flag, in case we have other things we want to use when developing, such as different logging levels or formatting.

File: parking/backend/__main__.py [hunk 2]
```diff
@@ -1,13 +1,34 @@
...
+    args: argparse.Namespace = parser.parse_args()
+
+    print(args)
```
Comment (HarveyHunt) [inline parking/backend/__main__.py]: You should drop this

File: parking/backend/__main__.py (hunk 1 above)
Comment (HarveyHunt) [reply parking/backend/__main__.py]: If not, it's common to use hyphens between long command line args.

File: parking/backend/db/dbaccess.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+import parking.backend.db.sql_constants as c
+from parking.shared.rest_models import ParkingLot
+
+logger = logging.getLogger('backend')
+
+
+class DbAccess(object):
+    @classmethod
+    async def create(cls, destination: str, loop: AbstractEventLoop, init_tables: bool=False) -> 'DbAccess':
+        self = DbAccess()
```
Comment (HarveyHunt) [inline parking/backend/db/dbaccess.py]: I think it is more pythonic to have: ```self = cls()```

File: parking/backend/db/dbaccess.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        self.pool: asyncpg.pool.Pool = await asyncpg.create_pool(dsn=destination, loop=loop)
+        if init_tables:
+            await self.create_tables()
+        return self
+
+    async def create_tables(self) -> None:
```
Comment (HarveyHunt) [inline parking/backend/db/dbaccess.py]: Does self need type hinting?

File: parking/backend/__main__.py (hunk 1 above)
Comment (HarveyHunt) [inline parking/backend/__main__.py]: The full stops are ugly :(

File: parking/backend/db/sql_constants.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+PARKINGLOTS_CREATE_TABLE = """
+CREATE TABLE ParkingLots (
+	id 	            serial PRIMARY KEY,
+	name 	        text NOT NULL,
+	capacity        integer NOT NULL,
+	lat 	        float NOT NULL,
+	long	        float NOT NULL,
+	price		    float NOT NULL,
+	num_available   integer NOT NULL,
+	num_allocated 	integer NOT NULL
```
Comment (HarveyHunt) [inline parking/backend/db/sql_constants.py]: This indentation is weird

File: parking/backend/db/sql_constants.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+);
+"""
+
+ALLOCATIONS_CREATE_TABLE = """
+CREATE TABLE Allocations (
+    user_id text NOT NULL,
+    park_id integer NOT NULL,
+
+    PRIMARY KEY(user_id),
```
Comment (HarveyHunt) [inline parking/backend/db/sql_constants.py:25]: Can the column be marked as a primary key in one line, as you did for ParkingLots.id?

File: tests/test_db.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+    new_price = 5
+
+    with testing.postgresql.Postgresql() as postgresql:
+        db = await DbAccess.create(postgresql.url(), event_loop, init_tables=True)
+
+        parking_lot = ParkingLot(100, 'test_name', 0.0, Location(0.0, 1.0))
+        park_id: id = await db.insert_parking_lot(parking_lot)
+        park_id: id = await db.update_parking_lot_availability(park_id, new_availability)
+        assert park_id == 1
+        park_id: id = await db.update_parking_lot_price(park_id, new_price)
```
Comment (HarveyHunt) [inline tests/test_db.py]: These type hints are wrong, shouldn't they be ```Optional[int]```?

File: tests/test_db.py [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+        park_id: id = await db.insert_parking_lot(parking_lot)
+        assert park_id == 1
+
+        async with db.pool.acquire() as conn:
+            lst: List[Record] = await conn.fetch('SELECT * from ParkingLots;')
+            assert len(lst) == 1
+
+
+@pytest.mark.asyncio
+async def test_delete_parking_lot(event_loop):
```
Comment (HarveyHunt) [inline tests/test_db.py:45]: Maybe add a test for deleting or updating a ParkingLot that doesn't exist, to make sure that the returned ID is None?

**Review by HarveyHunt — CHANGES_REQUESTED** [review]

File: parking/backend/__main__.py (hunk 1 above)
Comment (manfredlift) [reply parking/backend/__main__.py]: I disagree about the generic `--debug` flag. Running in debug mode should not mean running on a temporary database by default.

We could add a `--debug` flag for logging or other options later.

Btw, I added an option to `--reset-tables`, which resets the tables in a non-temporary database. This is necessary to clean any leftover data between runs of our backend.

Comment (Chippiewill) [reply parking/backend/__main__.py]: I would agree with Manfred here. Although I would rename ``--test_db`` to ``--temp_db`` since that better describes its purpose.

File: parking/backend/db/dbaccess.py (hunk 3 above)
Comment (manfredlift) [reply parking/backend/db/dbaccess.py]: It broke when I replaced it with your suggestion.

File: tests/test_db.py (hunk 7 above)
Comment (manfredlift) [reply tests/test_db.py]: True.

File: tests/test_db.py (hunk 8 above)
Comment (manfredlift) [reply tests/test_db.py:45]: Done.

**Comment by manfredlift** [comment]
Fixed stuff. Please have another look.

File: tests/test_db.py [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+import testing.postgresql
+from asyncpg import Record
+
+from parking.backend.db.dbaccess import DbAccess
+from parking.shared.location import Location
+from parking.shared.rest_models import ParkingLot
+
+
+@pytest.mark.asyncio
+async def test_empty_parking_lots(event_loop: AbstractEventLoop):
```
Comment (bensimner) [inline tests/test_db.py]: There should probably be an equivalent test for checking `create` creates an empty allocation table, too.

File: tests/test_db.py [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+
+@pytest.mark.asyncio
+async def test_update_parking_lot_price_and_availability_unsuccessful(event_loop):
+    new_availability = 10
+    new_price = 5
+
+    with testing.postgresql.Postgresql() as postgresql:
+        db = await DbAccess.create(postgresql.url(), event_loop, init_tables=True)
+
+        parking_lot = ParkingLot(100, 'test_name', 0.0, Location(0.0, 1.0))
```
Comment (bensimner) [inline tests/test_db.py]: Do we do any checking that we don't insert two parking lots into the same location?

If so, that should become a test, too.

Comment (bensimner) [reply tests/test_db.py]: And indeed anything else that could be duplicated but disallowed (parking lot names?)

Comment (manfredlift) [reply tests/test_db.py]: We don't do any checks like that as any new parking lots should be created by our 'admins' and shouldn't change that often. I'd move on from it now. Pls.

Comment (bensimner) [reply tests/test_db.py]: Move on from what?

Comment (manfredlift) [reply tests/test_db.py]: As in let's not do it now, please. :) This PR already blocks Harvey and me from working on other stuff.

Comment (bensimner) [reply tests/test_db.py]: If it's blocking you and seems to be working then merge it.

Comment (manfredlift) [reply tests/test_db.py]: @HarveyHunt merge when you've had another look. (does not need to be tonight).

**Comment by HarveyHunt** [comment]
Just a quick thought before I merge this, perhaps the tests should be wrapped in a single: ```with testing.postgresql.Postgresql() as postgresql:``` as otherwise a new SQL server is launched for each test - hurting test performance.

On my laptop, testing the DB stuff takes ~15 seconds which is annoying

**Comment by Chippiewill** [comment]
I say just merge it and create an issue to fix it

**Comment by manfredlift** [comment]
Technically, all unit tests should be independent and not reliant on the order they run in. This seems difficult to achieve with a shared database.

**Comment by HarveyHunt** [comment]
I'll merge this tomorrow, it conflicts with the .travis.yml and requirements.txt that I just merged. I tried to do it manually but there is another merge conflict with \_\_main\_\_.py and the git history here is weird -  Don't github whilst tired....

@manfredlift Can you rebase and squash into a single commit please? I'd rather be lazy and just use the nice green button on GH. :+1:

**Comment by HarveyHunt** [comment]
@manfredlift We could just use functions to drop and recreate the tables before each test runs, that'd be much quicker than spinning up a new server

**Comment by manfredlift** [comment]
Morning @HarveyHunt  👋. Rebased master + optimised tests like requested. Feel free to Squash and Merge whenever you are ready. 💯

**Comment by HarveyHunt** [comment]
Perfect, nice work :+1:

**Review by HarveyHunt — APPROVED** [review]
